### PR TITLE
Created new controller for switching modes for hardware interfaces

### PR DESCRIPTION
**Pull request by davetcoleman**
This is a generic solution for the [wiki page](https://github.com/ros-controls/ros_control/wiki/control_mode) on switching controllers. It is in response to issue https://github.com/ros-controls/ros_control/issues/99

### CODE REVIEW

**Comment by adolfo-rt** [comment]
I see that you have chosen a different approach from the one proposed in the [control_mode wiki page](https://github.com/ros-controls/ros_control/wiki/control_mode). Could you elaborate a bit on the pros and cons of both approaches?. Some questions that come to mind:

How would this controller be used to actually switch hardware control modes?. As far as I can see an enum value is being set, but no actual switching is done. How does this enum propagate to the hardware interfaces exposed by RobotHW?, or to the transmission propagation methods (when they exist)?. Without having given this much though, it seems to me that switching control modes can be very robot-specific.

It would seem that this proposal is for usecases in which the control mode does not change during runtime, as it's getting set at controller load-time. I would see a lot of value in allowing to change modes at runtime, as suggested in the [control_mode wiki page](https://github.com/ros-controls/ros_control/wiki/control_mode).

**Comment by davetcoleman** [comment]
I tried to follow @sachinchitta's wiki page's explanation to the best of my understanding, I just named the "switch" function "setMode" instead.

You can change control modes during runtime by simply stopping the, for example, velocity_mode_controller and loading a position_mode_controller. This controller is mode-agnostic and that [value is set in the param server](https://github.com/ros-controls/ros_controllers/blob/joint_mode_controller/joint_mode_controller/src/joint_mode_controller.cpp#L73). Its up to your hardware interface to check the `int*` provided by the `joint_mode_hardware_interface` every write loop and use the appropriate control method. You can see it implemented for Baxter [in the write() command](https://github.com/davetcoleman/baxter/blob/velocity_interface/baxter_control/src/arm_hardware_interface.cpp#L224) and in the [hardware interface constructor](https://github.com/davetcoleman/baxter/blob/velocity_interface/baxter_control/src/baxter_hardware_interface.cpp#L61).

I assume you saw the new hardware interface I added in ros_control?

I haven't thought about transmissions, but I think it would work too.

**Comment by davetcoleman** [comment]
I documented this new functionality (and moved the old documentation from the github wiki) to [ros.org](http://wiki.ros.org/ros_control/Tutorials/Switching%20control%20modes%20in%20hardware)

**Comment by adolfo-rt** [comment]
The Baxter example sets the mode in the hardware interface constructor, so it seems to be a pretty static configuration in which the mode remains unchanged during operation. I still wonder about the implications of changing mode at runtime.

The upside of doing a mode switch when starting the controller (as opposed to loading it) is that you can do more than one mode switch atomically through `controller_manager/switch _controllers`. You can only load one controller at a time.

I know this is implementation-specific, but still, some guidelines would be nice:
- Should inactive interfaces be removed from a `RobotHW` instance?, otherwise you'll be able to get resources which are currently disconnected from hardware. It seems odd to not get any errors and yet don't see the effect of a supposedly running controller.
- If we remove inactive interfaces from a `RobotHW` instance, what should we do with controllers previously running those interfaces, stop them?.

**Comment by davetcoleman** [comment]
> The Baxter example sets the mode in the hardware interface constructor

Yes, I have it default to position mode in the constructor, but this can be changed during run time by loading the [velocity_joint_mode_controller](https://github.com/davetcoleman/baxter/blob/velocity_interface/baxter_control/config/hardware_controllers.yaml#L11) which is an instance of `joint_mode_controller.cpp`. This will then set that `int` value to the velocity mode, or whatever other mode you want.

To change modes, you have to stop one mode controller and start the one of the mode you want to switch to, such as `position_joint_mode_controller`.

I'm not sure I completely understand how inactive interfaces work, resources, etc. I likely have not implemented this the best way in regards to that, but the basic functionality works for me.

**Comment by davetcoleman** [comment]
Ok, I spent more time testing different use cases and changed the mode controller so that it sets the mode `int` in the `starting()` function rather than the `init()` function. This allows it to switch during run time.

File: joint_mode_controller/src/joint_mode_controller.cpp [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+  }
+
+  void update(const ros::Time& time, const ros::Duration& period)
+  {}
+
+  void starting(const ros::Time& time) 
+  {
+    mode_handle_->setMode(joint_mode_);
+
+    ROS_INFO_STREAM_NAMED("init","Set joint mode to " << mode_handle_->getModeName(joint_mode_)
```
Comment (adolfo-rt) [inline joint_mode_controller/src/joint_mode_controller.cpp]: Log from realtime!

**Comment by adolfo-rt** [comment]
I'm not sure I understand the parameter specification of the interface. How would an example yaml file look like?. Is it possible to change mode of a set of joints from a single instance of this controller?.

I'm still unclear about what happens with inactive interfaces and controllers owning those resources. Maybe the project founders already have some practical experience with this, any thoughts @sachinchitta, @wmeeusse?.

**Comment by wmeeusse** [comment]
I have to dig a little deeper to understand all the details here, but I think there's still a component missing to deal with the resource management.  There's a way to prevent e.g. a position controller to be active while the robot is in velocity mode.

File: joint_mode_controller/src/joint_mode_controller.cpp (hunk 1 above)
Comment (davetcoleman) [inline joint_mode_controller/src/joint_mode_controller.cpp]: Fixed, thanks!

**Comment by davetcoleman** [comment]
>  There's a way to prevent e.g. a position controller to be active while the robot is in velocity mode.

Interesting, I didn't realize this. My implementation certainly does not prevent that. Can you provide some more pointers on how to do that?

**Comment by wmeeusse** [comment]
@davetcoleman The missing piece here is the check in the RobotHW class to see if there's a conflict between the set of controllers that is running. So e.g. you shouldn't be able to run a controller that uses the position interface when you are in velocity mode. This check happens in the "virtual bool checkForConflict(const std::list<ControllerInfo>& info) const" method in the RobotHW class.  With this PR it looks like we're providing people with a set of default tools to deal with mode switches, so we should make sure we also provide the resource checking. I think there might be an option to extend the default implementation in the base class to work with modes.

Can we extend this concept beyond the set of position/velocity/effort interfaces?  Could a 'mode' mean that you're restricting yourself to using one hardware interface at a time? In that case, the (string) parameter to the mode controller could be the name of the hardware interface that is allowed to be used when that controller is running (e.g. "PositionJointInterface"). Following the pattern in the current implementation, in the init method of the mode controller you can get a handle from the mode interface, which will verify that your mode string is a supported string. And then in the starting method of the mode controller you call enableMode() on the mode handle without any parameters.  Then, if you're still following here, the checkForConflict could check if there is a mode controller running, and if that's the case, it would make sure that all other controller are using the hardware interface that is allowed by the current mode. 

For a moment I was thinking that this would break down with the JointStateInterface, which can be in use in parallel with all other interfaces. But then I realized that the JointStateInterface does not claim any resources, since it's a read-only interface.

**Comment by davetcoleman** [comment]
> Can we extend this concept beyond the set of position/velocity/effort interfaces?

I provided those three and a "other" mode for future expansion. We can add others as needed.

> Could a 'mode' mean that you're restricting yourself to using one hardware interface at a time?

In my robot's case, yes

> Then, if you're still following here...

I kind of follow you, but I really just implemented the explanation given by Sachin on the Github wiki on how to do mode switching, and I figured I'd share it with others. I'm about to be out of the office for the month so don't have time to implement the rest of this, but I think we should go ahead and merge these 2 pull requests and add the user error checking later / you do it if you have time.

Thanks!

**Comment by wmeeusse** [comment]
> > Can we extend this concept beyond the set of position/velocity/effort interfaces?
> 
> I provided those three and a "other" mode for future expansion. We can add others as needed.

I think we can implement this using strings that point to interface types without having to enumerate the possible modes in the mode switching interface. That would make this tool usable for any custom interface that people are using.

> > Could a 'mode' mean that you're restricting yourself to using one hardware interface at a time?
> 
> In my robot's case, yes

Cool!  Having the tooling in place to allow user to limit their robot to use one interface at a time seems useful.

>  I really just implemented the explanation given by Sachin on the Github wiki on how to do mode switching, and I figured I'd share it with others. 

Definitely much appreciated! 

> I'm about to be out of the office for the month so don't have time to implement the rest of this, but I think we should go ahead and merge these 2 pull requests and add the user error checking later / you do it if you have time.

I would really like to stay very picky on what makes it into the default branch of ros_control.  My experience is that merging something in with the plan to fix/improve it later --even with the best intentions-- ends up straying in the repository for ever and has to be maintained and supported.  The place I'd like to get to is:
- Tooling that supports limiting the controller manager to one single interface at a time
- Support for any type of interface
- Resource conflict checking to guarantee that the user is not mixing interfaces when using the single interface mode.

**Comment by adolfo-rt** [comment]
The `checkForConflict(...)` implementation + the mode switching controller combination should be such that the following case is well supported:
1. `foo_joint` is in position mode, and `pos_controller` is using it (the resource is claimed).
2. We first attempt to start a mode switching controller that would switch `foo_joint` to velocity mode.
3. The switch fails, because `foo_joint` resource has not been released by `pos_controller`.
4. We now attempt to switch `foo_joint` to velocity mode _and_ stop `foo_controller`. The switch now succeeds.

This would effectively prevent us from having dangling controllers that are writing to a disconnected interface.

**Comment by fmessmer** [comment]
I tried to split the various features in separate pull requests:
#108 velocity_controller/JointPositionController
#109 velocity interface for JointTrajectoryController
#110 joint_mode_controller

**Comment by davetcoleman** [comment]
I believe this has been merged into indigo in various commits and can be closed now. Please re-open if I am mistaken, I'm just cleaning up my pending pull requests.

**Comment by henriquejsfj** [comment]
Hello,

Was this PR merged anywhere? The links are broken (even docs one), the branches deleted, I didn't find anything about joint_mode_controller at Google...

I would like this feature of switching the interface at runtime in the write function. Actually, the Gazebo requires a sim_control_mode as used in https://github.com/PickNikRobotics/ros_control_boilerplate. Although in the Gazebo case this is set at the constructor and doesn't change at runtime.

In my case, I have a brushed DC Motor. The (mathematical) correct way to control it should be actuating at torque (effort), but everybody usually actuates at velocity (which also works). Thus, I would like to use both and when switching the controller (as this guy https://answers.ros.org/question/259022/switching-between-controllers-with-ros_control-controller_manager/) my hardware interface would use the proper interface (Velocity or Effort).

Thank you in advance.

**Pull request closed without merging**